test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/AttributeRouteTest.cs

```diff
@@ -118,5 +103,513 @@ public class AttributeRouteTest
+                    Assert.Equal(RoutePrecedence.ComputeMatched(e.RouteTemplate), e.Precedence);
+                    Assert.Equal("BLOG_INDEX", e.RouteName);
+                    Assert.Equal("api/Blog/{id}", e.RouteTemplate.TemplateText);
+                    Assert.Same(handler, e.Target);
+                    Assert.Collection(
+                        e.TemplateMatcher.Defaults.OrderBy(kvp => kvp.Key),
+                        kvp => Assert.Equal(new KeyValuePair<string, object>(TreeRouter.RouteGroupKey, "1"), kvp));
+                    Assert.Same(e.RouteTemplate, e.TemplateMatcher.Template);
+                });
+        }
+
+        [Fact]
+        public void AttributeRoute_GetEntries_CreatesMatchingEntry_WithConstraint()
+        {
+            // Arrange
+            var actions = new List<ActionDescriptor>()
+            {
+                new ActionDescriptor()
+                {
+                    AttributeRouteInfo = new AttributeRouteInfo()
+                    {
+                        Template = "api/Blog/{id:int}",
+                        Name = "BLOG_INDEX",
+                        Order = 17,
+                    },
+                    RouteConstraints = new List<RouteDataActionConstraint>()
+                    {
+                        new RouteDataActionConstraint(TreeRouter.RouteGroupKey, "1"),
+                    },
+                    RouteValueDefaults = new Dictionary<string, object>()
+                    {
+                        { "controller", "Blog" },
+                        { "action", "Index" },
+                    },
+                },
+            };
+
+            var actionDescriptorProvider = CreateActionDescriptorProvider(actions);
+
+            var handler = CreateHandler().Object;
+            var route = CreateRoute(handler, actionDescriptorProvider.Object);
+
+            // Act
+            var entries = route.GetEntries(actionDescriptorProvider.Object.ActionDescriptors);
+
+            // Assert
+            Assert.Collection(
+                entries.MatchingEntries,
+                e =>
+                {
+                    Assert.Single(e.Constraints, kvp => kvp.Key == "id");
+                    Assert.Equal(17, e.Order);
+                    Assert.Equal(RoutePrecedence.ComputeMatched(e.RouteTemplate), e.Precedence);
+                    Assert.Equal("BLOG_INDEX", e.RouteName);
+                    Assert.Equal("api/Blog/{id:int}", e.RouteTemplate.TemplateText);
+                    Assert.Same(handler, e.Target);
+                    Assert.Collection(
+                        e.TemplateMatcher.Defaults.OrderBy(kvp => kvp.Key),
+                        kvp => Assert.Equal(new KeyValuePair<string, object>(TreeRouter.RouteGroupKey, "1"), kvp));
+                    Assert.Same(e.RouteTemplate, e.TemplateMatcher.Template);
+                });
+        }
+
+        [Fact]
+        public void AttributeRoute_GetEntries_CreatesMatchingEntry_WithDefault()
```

Comment:
This is the test that verifies the fix (along with the tweaked functional test). There are tests for defaults in attribute routes, but they life in the routing repo. What was missing here was the correct wire-up of values.